plugins/rapid7_insightvm/komand_rapid7_insightvm/triggers/scan_completion/trigger.py

```diff
@@ -0,0 +1,274 @@
+                    return scan.get("id")
+        else:
+            response = resource_helper.resource_request(endpoint=endpoint, method="get", params={"sort": "id,desc"})
+
+            for scan in response.get('resources'):
+                if scan.get("status") == "finished":
+                    self.logger.info(f"Latest finished scan ID: {scan.get('id')}")
+                    return scan.get("id")
+
+class ScanQueries:
+    @staticmethod
+    def query_results_from_latest_scan(scan_id: int) -> str:
+        """
+        Generate an SQL query string needed to to retrieve all the necessary outputs
+
+        :param scan_id: Scan ID to query against
+        :return: The completed query string
+        """
+
+        return (
+            f"SELECT fasvi.scan_id, fasvi.asset_id, fasvi.vulnerability_id, dvr.source, daga.asset_group_id, da.host_name, da.ip_address, dss.solution_id, dss.summary, dv.nexpose_id "  # nosec B608
+            f"FROM fact_asset_scan_vulnerability_instance AS fasvi "
+            f"JOIN dim_asset_group_asset AS daga ON (fasvi.asset_id = daga.asset_id) "
+            f"JOIN dim_asset AS da ON (fasvi.asset_id = da.asset_id) "
+            f"JOIN dim_vulnerability AS dv ON (fasvi.vulnerability_id = dv.vulnerability_id) "
+            f"JOIN dim_vulnerability_reference AS dvr ON (fasvi.vulnerability_id = dvr.vulnerability_id) "
+            f"JOIN dim_solution AS dss ON (dv.nexpose_id = dss.nexpose_id) "
+            f"WHERE fasvi.scan_id = {scan_id} "
+        )
+
+
+class Cache:
+    @staticmethod
+    def get_cache_site_scans() -> dict:
+        try:
+            return json.loads(util.read_from_cache(ScanCompletion.CACHE_FILE_NAME))
+        except ValueError as error:
+            raise PluginException(cause="Failed to load cache file", assistance=f"Exception returned was {error}")
+
+
+class Util:
+    @staticmethod
+    def filter_results(params: dict, csv_row: dict):
+        """
+        Filter the outputted results based on the user inputs.
+
+        :param params: Input params
+        :param csv_row: Dict row of the csv results
+
+        :return: New object containing only the necessary fields for the required
+        output.
+        """
+        # Input retrieval
+        asset_group = params.get(Input.ASSET_GROUP, None)
+        cve = params.get(Input.CVE, None)
+        hostname = params.get(Input.HOSTNAME, None)
+        source = params.get(Input.SOURCE, None)
+        ip_address = params.get(Input.IP_ADDRESS, None)
+
+        new_dct = {
+            "asset_id": int(csv_row["asset_id"]),
+            "ip_address": csv_row["ip_address"],
+            "hostname": csv_row["host_name"],
+            "vulnerability_id": csv_row["vulnerability_id"],
+            "nexpose_id": csv_row["nexpose_id"],
+            "solution_id": Util.strip_msft_id(csv_row.get("solution_id", "")),
+            "solution_summary": csv_row["summary"],
+        }
```

Comment:
Is there any possibility to get `KeyError`? 😄

Reply:
Theres not but it can't hurt to include a `try: except` anyway 😄